File: synth/synth/mutators/missing_fields.py
"""Pre-render mutators that drop specific fields from the spec and align ground truth."""

from __future__ import annotations
import random


def drop_invoice_number(spec, gt, rng: random.Random) -> tuple:
    """Remove invoice number from both the rendered document and ground truth."""
    spec.invoice_number = ""
    gt.invoice_number = None
    gt._meta.missing.append("invoice_number")
    gt._meta.mutations.append("missing_invoice_number")
    return spec, gt


def drop_due_date(spec, gt, rng: random.Random) -> tuple:
    spec.due_date = None
    gt.due_date = None
    gt._meta.missing.append("due_date")
    gt._meta.mutations.append("missing_due_date")
    return spec, gt


def drop_supplier_cif(spec, gt, rng: random.Random) -> tuple:
    """Render supplier block without CIF (field missing from document)."""
    spec.supplier.cif = ""
    gt._meta.missing.append("supplier_cif")
    gt._meta.mutations.append("missing_cif_supplier")
    return spec, gt


def drop_total(spec, gt, rng: random.Random) -> tuple:
    """For albaranes — mark as no-total delivery note."""
    gt.total_amount = None
    gt._meta.missing.append("total")
    gt._meta.mutations.append("missing_total")
    return spec, gt


def drop_date(spec, gt, rng: random.Random) -> tuple:
    spec.invoice_date = ""
    gt.invoice_date = None
    gt._meta.missing.append("invoice_date")
    gt._meta.mutations.append("missing_date")
    return spec, gt


def add_negative_line(spec, gt, rng: random.Random) -> tuple:
    """Add a product return line (negative quantity)."""
    if spec.items:
        return_item = dict(spec.items[0])
        return_item["quantity"] = -abs(return_item["quantity"])
        return_item["total_price"] = round(return_item["unit_price"] * return_item["quantity"], 2)
        spec.items.append(return_item)
        from synth.schema.ground_truth import LineItem
        gt.line_items.append(LineItem(
            description=return_item["description"],
            quantity=return_item["quantity"],
            unit=return_item["unit"],
            unit_price=return_item["unit_price"],
            total_price=return_item["total_price"],
        ))
        gt._meta.mutations.append("negative_quantity")
    return spec, gt


MUTATOR_MAP = {
    "missing_invoice_number": drop_invoice_number,
    "missing_due_date":       drop_due_date,
    "missing_cif_supplier":   drop_supplier_cif,
    "missing_total":          drop_total,
    "missing_date":           drop_date,
    "negative_quantity":      add_negative_line,
}

File: synth/js/mutators/missing_fields.mjs
export function dropInvoiceNumber(spec, gt) {
  spec.invoice_number = '';
  gt.invoice_number = null;
  gt._meta.missing.push('invoice_number');
  gt._meta.mutations.push('missing_invoice_number');
}

export function dropDueDate(spec, gt) {
  spec.due_date = null;
  gt.due_date = null;
  gt._meta.missing.push('due_date');
  gt._meta.mutations.push('missing_due_date');
}

export function dropSupplierCif(spec, gt) {
  spec.supplier.cif = '';
  gt._meta.missing.push('supplier_cif');
  gt._meta.mutations.push('missing_cif_supplier');
}

export function dropTotal(spec, gt) {
  gt.total_amount = null;
  gt._meta.missing.push('total');
  gt._meta.mutations.push('missing_total');
}

export function dropDate(spec, gt) {
  spec.invoice_date = '';
  gt.invoice_date = null;
  gt._meta.missing.push('invoice_date');
  gt._meta.mutations.push('missing_date');
}

export function addNegativeLine(spec, gt, rng) {
  if (!spec.items.length) return;
  const src = { ...spec.items[0] };
  src.quantity = -Math.abs(src.quantity);
  src.total_price = Math.round(src.unit_price * src.quantity * 100) / 100;
  spec.items.push(src);
  gt.line_items.push({
    description: src.description,
    quantity: src.quantity,
    unit: src.unit,
    unit_price: src.unit_price,
    total_price: src.total_price,
  });
  gt._meta.mutations.push('negative_quantity');
}

export const MUTATOR_MAP = {
  missing_invoice_number: (s, g)       => dropInvoiceNumber(s, g),
  missing_due_date:       (s, g)       => dropDueDate(s, g),
  missing_cif_supplier:   (s, g)       => dropSupplierCif(s, g),
  missing_total:          (s, g)       => dropTotal(s, g),
  missing_date:           (s, g)       => dropDate(s, g),
  negative_quantity:      (s, g, rng)  => addNegativeLine(s, g, rng),
};
